### DISCUSSION TITLE
Powertools for AWS Lambda (TypeScript) v3 - Ideas and notes

### DISCUSSION THREAD

**Discussion by dreamorosi** [aws-powertools/powertools-lambda-typescript#2560] [Ideas]
I would like to open a discussion to gather and ideally discuss ideas and changes for our next v3 milestone.

Note that at the time of writing we don't have any date nor plans for our next major version, however I think it'd be good to start gathering notes & ideas for things we might encounter along the way that we wish we could address but can't due to major refactors or breaking changes.

**Comment by dreamorosi** [comment]
Relating to the Batch Processor utility, the components of the utility follow an inheritance pattern with some abstract classes and methods. During the initial implementation we weren't necessarily the most diligent and so almost all methods were marked as `public`.

Some of these methods (my feeling is that it's most of them) could potentially be converted to `protected` instead, since they are meant to be used only by classes extending the base class.

As part of v3 we could consider doing an exercise of minor refactoring and clean up the public API.

**Comment by HaaLeo** [comment]
Maybe consider #749 for v3 if that would cause any breaking changes.

**Comment by dreamorosi** [comment]
In aws-powertools/powertools-lambda-python#4907 the `extractDataFromEnvelope()` function part of the JMESPath utility was renamed to simply `query()`, we should do the same.

**Comment by dreamorosi** [comment]
As discussed in #3085, in v3 we will switch the `experimentalDecorators` setting in the `tsconfig.json` file to `false`.

In v5.0 of TypeScript, the default changed to Stage 3 decorators, which modify the signature of the decorator significantly as discussed more in depth [here](https://github.com/aws-powertools/powertools-lambda-typescript/issues/1375#issuecomment-1497722764).

While doing this, we should also make our decorators compatible with asynchronous methods only. Today they support both but in reality they implicitly convert sync methods to async under the hood. This is not a huge issue when decorating the main function handler since AWS Lambda will handle the promise correctly, but when decorating arbitrary methods this can cause subtle bugs.

**Reply by codyfrisch** [reply]
> As discussed in #3085, in v3 we will switch the `experimentalDecorators` setting in the `tsconfig.json` file to `false`.
> 
> In v5.0 of TypeScript, the default changed to Stage 3 decorators, which modify the signature of the decorator significantly as discussed more in depth [here](https://github.com/aws-powertools/powertools-lambda-typescript/issues/1375#issuecomment-1497722764).
> 
> While doing this, we should also make our decorators compatible with asynchronous methods only. Today they support both but in reality they implicitly convert sync methods to async under the hood. This is not a huge issue when decorating the main function handler since AWS Lambda will handle the promise correctly, but when decorating arbitrary methods this can cause subtle bugs.

As long as this has no impacts on projects that still depend on the old experimental decorators - inversify comes to mind. Just something to consider. (They depend on parameter decorators which aren't in the stage 3 spec). I don't know if offering both is even viable.

**Reply by dreamorosi** [reply]
I have yet to do an in depth PoC, but so far I'm with you on it not being viable.

In terms of impact, we're leaving this for a major version on purpose since it will probably result in a breaking change.

For what it's worth, we don't yet have any concrete plans for v3 and I don't expect us to start working on it before end of the year at the very earliest, and anyway after having cleared [our current roadmap](https://docs.powertools.aws.dev/lambda/typescript/latest/roadmap/) for the year.

As soon as v3 plans are more concrete I'll publish announcements in the repo and other channels to communicate changes like this one well in advance like we did for v2, so that customers can prepare. Even after v3 is released, we will still support the current version for 6 extra months according to [our versioning policy](https://docs.powertools.aws.dev/lambda/typescript/latest/versioning/).

This is to say that there's still a ton of time before this becomes a problem, however I think it's also important to stay up to date with the ecosystem and acknowledge that there's demand for moving to the new implementation which at least for now looks like it's going to be the final one.

**Comment by dreamorosi** [comment]
Following the implementation of #3160 that adds async processing to SQS FIFO queues, to avoid breaking changes we had to invert the convention of `XXProcessor` being async and `XXProcessorSync` being the synchronous opt-in. In this case the existing SQS FIFO processor stayed sync and we added an asynchronous one.

At minimum, on v3 we should standardize this implementation to follow the rest of the Batch Processor classes.

Additionally, as @am29d pointed out, there might be an opportunity to streamline the implementation and API of this utility significantly while making minimal breaking changes.

For example, we could explore reducing the number of classes to have only one batch processor for event source, and then have both synchronous and asynchronous methods on it, so that customers don't have to think which one they need to import and use but just call `BatchProcessor.process()` (async) or `BatchProcessor.processSync()`.

We are unsure if this is possible or the extent of breaking changes it'll require - for now we're adding this to the v3 list to explore further.

**Comment by dreamorosi** [comment]
Another item we're looking to remove or rework is the custom configuration services from the core utilities.

Today it's possible to pass a custom configuration service to the constructor of Tracer, Metrics, and Logger which will be used during initialization.

Currently this feature doesn't seem to have gotten any traction, most likely also because it's not documented anywhere, but also because it's too rigid. 

Defining your custom configurations service requires you to reimplement the entire interface, which might not be desirable when you want to only override one behavior.

**Comment by dreamorosi** [comment]
In v3 we should stop including the `service` dimension to metrics emitted when the service name is not specified via `POWERTOOLS_SERVICE_NAME` env variable or `serviceName` constructor options.

Currently we still add the dimension but with value `service_undefined`, which adds no value and only increases the size of the log emitted.

Customers have asked us to change this but doing so requires a breaking change, so I'm adding this here.

More info in #3374.

**Comment by am29d** [comment]
Change `POWERTOOLS_TRACE_ENABLED` to `POWERTOOLS_TRACE_DISABLED` to make it consistent across all runtimes.

**Comment by dreamorosi** [comment]
As a customer, when retrieving a SSM parameter stored as `StringList` and applying the `transform: 'json'` I want to get back an array of strings rather than a string like this `foo,bar`.

Currently we do the latter, but we should consider doing the former even though the `StringList` is not a proper JSON string.

More info in #2041.

**Comment by dreamorosi** [comment]
Following the discussion in #3781, we could consider changing how the partition and sort key are created in Idempotency.

When using composite keys, our current implementation sets the keys like this:
- PK: `idempotency` (or other static value)
- SK: `function-name#hash-value` (the hash is derived from the payload)

This is done to allow us to read/write items while knowing only the payload and thus minimize reads.

The problem with this design - which again, affects only customers using composite keys - is that it creates a [hot partition](https://aws.amazon.com/blogs/database/part-2-scaling-dynamodb-how-partitions-hot-keys-and-split-for-heat-impact-performance/) due to the static nature of the PK.

For customers who are seeing hot partition issues and are also sharing the same DynamoDB table for multiple idempotent functions, the issue can be partially be mitigated by using the `keyPrefix` option documented [here](https://docs.powertools.aws.dev/lambda/typescript/latest/utilities/idempotency/#customizing-the-idempotency-key-prefix)

If instead hot partition becomes an issue with a single function - then there's nothing else to do other than extend the `DynamoDBPersistenceLayer` class and update the `getKey()` method to create a different PK value.

This entire experience is not optimal, and the discussion in the linked issue highlighted that there might be a better way, which consists in writing the keys like this (the payload hash is used in both keys):
- PK `idempotency#hash-value`
 SK: `function-name#hash-value`

This is a clever design that we did not think of, and we should make extensive tests before deciding to implement it. However given that it's a breaking change in how the keys are generated, we're putting this on hold until our next major version to reconsider.

**Comment by svozza** [comment]
Couple of ideas:

1. I am strongly in favour of moving to ESM only.
2. Batch utilities could be refactored to the more functional approach that the Kafka consumer follows. The current implementation has quite a lot of internal mutable state that could be simplified imo.

**Comment by dreamorosi** [comment]
After experimenting with Microsoft's native TypeScript compiler in PR #4923, we observed significant performance improvements in build times (5-6x faster locally, 33% faster in CI). However, one of the main drawbacks was that the native compiler required less strictly-typed decorator implementations (using `as any` casts).

Since this discussion touches on decorator patterns and given that new decorators are likely part of future v3 considerations, we should revisit adopting the native TypeScript compiler when we start working on v3 - assuming it has also become more stable by then.

The performance benefits are substantial enough to warrant reconsideration once we address the decorator implementation concerns.

Reference: #4923

**Comment by svozza** [comment]
Following #5667 / #5668 in the HTTP event handler, an idea for v3 on how we decide `isBase64Encoded` for responses.

**Problem today.** `reqCtx.isBase64Encoded` does two jobs: it is the documented user knob to force base64, and it is also the framework's own memory that "the handler result was binary". Because the second use piggybacks on the first, the framework can only ever set it to `true`, and every later change to `reqCtx.res` (error handler, middleware short-circuit, compress) has to be careful not to ship a stale flag. #5668 patches the call sites, but the conflation is still there. On top of that, the fallback heuristic treats only `image/*`, `audio/*`, `video/*` as binary and everything else as text, so unknown binary types (`application/pdf`, `application/zip`, fonts) get corrupted through `response.text()`.

**Proposal.** Make the encoding a pure function of the final `Response` headers, decided once in `webResponseToProxyResult`, with no in-flight state:

- `content-encoding` present → base64
- text allowlist (`text/*`, `application/json`, `+json`, `application/xml`, `+xml`, `application/javascript`, `application/yaml`, `+yaml`) → text
- no `content-type` → text
- everything else → base64

For that to be sufficient, `handlerResultToWebResponse` has to leave truthful headers: a binary result with no content type gets `application/octet-stream` instead of today's `application/json`, and a proxy result with `isBase64Encoded: true` is decoded to bytes at the boundary and the flag forgotten. Then `reqCtx.isBase64Encoded` can be removed (the override becomes "set the content type"), and compress can read `reqCtx.isHttpStreaming` instead of sniffing the `transfer-encoding: chunked` header we plant on the initial response, which is the same smell.

**Escape hatch** for textual bodies under unknown types: a router option (`textContentTypes: ['application/x-ndjson']`) and/or a per-response header that we strip on the way out.

**Precedent.** `application/octet-stream` for a bare binary body is what Express, Fastify, and Koa do. The text-allowlist, binary-by-default edge rule is exactly what the Rust runtime's `lambda_http` (used by Lambda Web Adapter) and Hono's `aws-lambda` adapter implement; both also treat a missing content type as text. `lambda_http` ships a header escape hatch, Hono a predicate option. serverless-express and Powertools today are the two on the binary-allowlist side.

**Breaking for consumers:** unknown content types go out base64 encoded; `reqCtx.isBase64Encoded` is removed; bare binary results carry `application/octet-stream`. JSON/text/HTML handlers, `Response`-returning binary handlers, proxy results with `isBase64Encoded: true`, and streaming are unaffected. The trade-off is that a REST API without `binaryMediaTypes` serves base64 text for a textual body under an unknown type until it is allowlisted, which is loud rather than silent corruption.
